File: 21851094-沈吴越/Project01/OpenGL_First_Homework/OpenGL_First_Homework/Draw_Teapot.cpp
#include <gl/glut.h>
#include <Windows.h>
#include <algorithm>
#include <math.h>

static int window_width = 800, window_height = 800;
static int display_pos_x = (GetSystemMetrics(SM_CXSCREEN) - window_width) / 2;
static int display_pos_y = (GetSystemMetrics(SM_CYSCREEN) - window_height) / 2;
static bool mouseleftdown = false, mouserightdown = false, mousemiddledown = false;
static float mouse_x, mouse_y, camera_distance_x, camera_distance_y, camera_angle_x, camera_angle_y, size = 1.0;


void reshape(int width, int height)
{
	glViewport(0, 0, width, height);
	glMatrixMode(GL_PROJECTION);
	gluPerspective(80.0, (float)width / height, 1.0, 1000);
	glLoadIdentity();
}

void display()
{
	glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
	glMatrixMode(GL_MODELVIEW);
	glColor3f(1.0, 0.0, 0.0);
	glLoadIdentity();
	glScalef(size, size, size);
	glTranslatef(camera_distance_x, camera_distance_y, 0);
	glRotatef(camera_angle_x, 1, 0, 0);
	glRotatef(camera_angle_y, 0, 1, 0);
	glutWireTeapot(0.5);
	glutSwapBuffers();
}

void mouse_click(int button, int state, int x, int y)
{
	mouse_x = x;
	mouse_y = y;
	size = 1.0;
	if (button == GLUT_LEFT_BUTTON) {
		if (state == GLUT_DOWN) {
			mouseleftdown = true;
		}
		else {
			mouseleftdown = false;
		}
	}
	else if (button == GLUT_RIGHT_BUTTON) {
		if (state == GLUT_DOWN) {
			mouserightdown = true;
		}
		else {
			mouserightdown = false;
		}
	}
	else if (button == GLUT_MIDDLE_BUTTON) {
		if (state == GLUT_DOWN) {
			mousemiddledown = true;
		}
		else {
			mousemiddledown = false;
		}
	}
}

void mouse(int x, int y)
{
	if (mouseleftdown) {
		camera_angle_y += (x - mouse_x);
		camera_angle_x += (y - mouse_y);
		mouse_x = x;
		mouse_y = y;
	}
	if (mouserightdown) {
		camera_distance_x += (x - mouse_x)*0.002;
		camera_distance_y -= (y - mouse_y)*0.002;
		mouse_x = x;
		mouse_y = y;
	}
	if (mousemiddledown) {
		size += (mouse_y - y)*0.002;
		mouse_y = y;
	}
	glutPostRedisplay();
}


void init()
{
	glEnable(GL_DEPTH_TEST);
}


int main(int argc, char* argv[])
{
	glutInit(&argc, argv);
	glutInitDisplayMode(GLUT_DEPTH | GLUT_DOUBLE | GLUT_RGBA);
	glutInitWindowSize(window_width, window_height);
	glutInitWindowPosition(display_pos_x, display_pos_y);
	glutCreateWindow("OpenGL-第一次作业-沈吴越21851094-绘制茶壶");
	init();
	glutMouseFunc(mouse_click);
	glutMotionFunc(mouse);
	glutDisplayFunc(display);
	glutReshapeFunc(reshape);
	glutMainLoop();
	return 0;
}

File: 21851486邓砚庭/Project01/drawTeapot.py

__author__ = 'Dante'


import sys

from interaction import Interactor

try:
    from OpenGL.GLUT import *
    from OpenGL.GLU import *
    from OpenGL.GL import *
except:
    print('PyOpenGL not found!')
    sys.exit(1)

red = '\x1b[31m'
origin = '\x1b[0m'

print()
print(f'Press {red}ESC{origin} to close the window.')
print(f'Hold and Drag {red}Right Mouse Key{origin} to move.')
print(f'Hold and Drag {red}Left Mouse Key{origin} to rotate.')
print(f'Use {red}Mouse Wheel{origin} to scale, if not working, hold and drag the wheel.')


def display():
    '''Display a "real" teapot.'''
    glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT)
    glMatrixMode(GL_PROJECTION)
    glLoadIdentity()
    xSize, ySize = glutGet(GLUT_WINDOW_WIDTH), glutGet( GLUT_WINDOW_HEIGHT)
    gluPerspective(60, float(xSize) / float(ySize), 0.1, 50)
    glMatrixMode(GL_MODELVIEW)
    glLoadIdentity()
    glTranslatef(0, 0, -4)
    global interaction
    interaction.applyTransformation()
    glCallList(tkList)
    glDisable(GL_LIGHTING)
    glColor3f(9, 5, 0.5)
    glRasterPos3f(1.8, .5, 0)

    glEnable(GL_LIGHTING)
    glutSwapBuffers()

def init():
    glClearColor (0.99, 0.99, 0.99, 0)
    glEnable(GL_DEPTH_TEST)
    glShadeModel(GL_SMOOTH)

    

    # Teapot Apperance
    glMaterialfv(GL_FRONT_AND_BACK, GL_AMBIENT, [0.2125, 0.1275, 0.054, 1])
    glMaterialfv(GL_FRONT_AND_BACK, GL_DIFFUSE, [0.714, 0.4284, 0.18144, 1])
    glMaterialfv(GL_FRONT_AND_BACK, GL_SPECULAR, [0.393548, 0.271906, 0.16672, 1])
    glMaterialf(GL_FRONT_AND_BACK, GL_SHININESS, 25.6)

    # Initialise interaction
    global interaction
    interaction = Interactor(.01, 1)
    global tkList
    tkList = glGenLists(1)
    glNewList(tkList, GL_COMPILE)
    glutSolidTeapot(1.0)
    glEndList()

# Show it
glutInit(sys.argv)
glutInitDisplayMode(GLUT_DOUBLE | GLUT_RGB | GLUT_DEPTH)
glutInitWindowSize(500, 500)
glutCreateWindow('Real Teapot')
init()
interaction.registerCallbacks()
glutDisplayFunc(display)
glutMainLoop()
